File: Chapter-5-Classes-and-Objects/012-a-sample-program-that-shows-how-to-use-common-friend-function-to-exchange-the-private-values-of-two-classes.cpp
#include <iostream>

using namespace std;

class class_2;

class class_1
{
    int value;

public:
    void inputData(int a)
    {
        value = a;
    }
    void displayData()
    {
        cout << value << endl;
    }
    friend void exchangeValue(class_1 &, class_2 &);
};

class class_2
{
    int value;

public:
    void inputData(int a)
    {
        value = a;
    }
    void displayData()
    {
        cout << value << endl;
    }
    friend void exchangeValue(class_1 &, class_2 &);
};

void exchangeValue(class_1 &x, class_2 &y)
{
    int temp = x.value;
    x.value = y.value;
    y.value = temp;
}

int main()
{
    class_1 C1;
    class_2 C2;

    C1.inputData(100);
    C2.inputData(200);

    cout << "Values before exchange" << endl;
    C1.displayData();
    C2.displayData();

    exchangeValue(C1, C2); /// swapping

    cout << "Values after exchange" << endl;
    C1.displayData();
    C2.displayData();
}

File: Chapter-5-Classes-and-Objects/012-a-sample-program-that-shows-how-to-use-common-friend-function-to-exchange-the-private-values-of-two-classes.py
class Class2:
    def __init__(self):
        self.value = 0
    
    def input_data(self, a):
        self.value = a
    
    def display_data(self):
        print(self.value)
    
class Class1:
    def __init__(self):
        self.value = 0
    
    def input_data(self, a):
        self.value = a
    
    def display_data(self):
        print(self.value)

def exchange_value(x, y):
    temp = x.value
    x.value = y.value
    y.value = temp

if __name__ == '__main__':
    C1 = Class1()
    C2 = Class2()

    C1.input_data(100)
    C2.input_data(200)

    print("Values before exchange")
    C1.display_data()
    C2.display_data()

    exchange_value(C1, C2) # Swap the values

    print("Values after exchange")
    C1.display_data()
    C2.display_data()